README.md
# Mouse Trap Monitor

As featured in issue 108 of [The MagPi Magazine (August 2021)](https://magpi.raspberrypi.org/issues/108) 
and on the [Raspberry PI Blog](https://www.raspberrypi.org/blog/humane-mouse-trap-the-magpi-108).

A PIR Raspberry PI powered Python script to detect potential mouse movements.

## What is it's purpose?

If you have a humane mouse trap, it is especially important to know whether there is a mouse in it. If there is and you don't check, then it will eventually run out of food and water which would be bad news for the mouse...

Checking the traps are boring and it is useful to know whether there is likely to be activity in remote locations e.g. a garage. This project was designed to help with that problem.

## How does it work?

The PIR sensor needs to ideally point at the (preferably) humane mouse trap. The below values are based on the default example.

* The PIR sensor will look for movement every 0.1 seconds.
* When movement is detected, it will add one to the counter and will wait for 30 seconds.
* After 24 hours (approx) it will fire off a request to IFTTT which in turn will trigger a request to Pushbullet. You will then get a notification on your phone telling you how many instances of movement have been detected in the last 24 hours. The counter is reset.

    If the movement is significantly higher than prevous days e.g. 20% higher then there is a chance there might be a mouse in the trap and it may be worth checking out!

> Example: In my case I was getting approx 800ish triggers a day. At 1036 I found that there was a mouse in the trap. Note that there is a degree of variation and nothing is exact so use your instincts.

## Hardware

This is what I used for this project, but most Raspberry PI models would work just as well providing they have GPIO to connect the PIR sensor to:

* Raspberry PI Zero WH (Wifi for networking and header for easy GPIO use but your use case may vary!)
* PIR sensor
* 3 Dupont jumper wires to connect them with
* (Optional) Case for PI and a coffee lid to protect the sensor!

## Software

### Installation / Setup

> Note: This is not a thorough guide, although if you know roughly what you are doing it is probably enough

1. Follow the [PiHut IFTTT Tutorial](https://thepihut.com/blogs/raspberry-pi-tutorials/using-ifttt-with-the-raspberry-pi)
2. Install python3 if it isn't already (may be covered by IFTTT tutorial)
3. Install python-dotenv:

        pip3 install -U python-dotenv
4. Clone this repository into a directory of your choice e.g. your home directory at `~/`:

        git clone git@github.com:vectrexman/mouse-trap-monitor.git
5. From within your cloned directory e.g. `cd ~/mouse-trap-monitor` Copy the example `.env.example` file to `.env` and alter variables to your preference including putting in your IFTTT key *important*

        cp .env.example .env

    If you don't do this you will likely get errors as it won't know what values to use.

### Running the script

From the git clone'd project directory:

From anywhere (substituting the locaton as relevant):

    python3 ~/mouse-trap-monitor/start.py

Or from within the directory that you cloned `mouse-trap-monitor`:

    python3 ./start.py

### Dependencies

#### Origins

This project is built on the excellent work undertaken as part of these projects:

Based on [PiHut IFTTT Tutorial](https://thepihut.com/blogs/raspberry-pi-tutorials/using-ifttt-with-the-raspberry-pi) which 
is in turn based on [CamJam Edukit 2](https://github.com/CamJam-EduKit/EduKit2/blob/master/CamJam%20Edukit%202%20-%20RPi.GPIO/Code/5-PIR.py)

#### Python Package Dependencies

    pip3 install -U python-dotenv
        
(https://pypi.org/project/python-dotenv/)

## Appendix

Project formally known as GaragePI. Re-named it to make it more specific for other peoples potential (re)use!

CHANGELOG.md
# Mouse Trap Monitor Changelog

## [Unreleased]

### Added
- Added additonal exception handling including an additional timeout on failure

## [1.0.1] - 2022-02-12

### Added
- Should now attempt to re-try five times before giving up
- Specified connection timeouts

## [1.0.0] - 2021-06-21

### Added
- Initiate Release



garagepi.py
#! /usr/bin/python

# Based on https://cdn.shopify.com/s/files/1/0176/3274/files/iftttpir.py from the PiHut

# Imports
import RPi.GPIO as GPIO
import time
import requests
from dotenv import load_dotenv
import os
from datetime import datetime

# Set the GPIO naming convention
GPIO.setmode(GPIO.BCM)

# Turn off GPIO warnings
GPIO.setwarnings(False)

# Set a variable to hold the GPIO Pin identity
pinpir = 17

# Set GPIO pin as input
GPIO.setup(pinpir, GPIO.IN)

# Variables to hold the current and last states
currentstate = 0
previousstate = 0

# Load Environment Variables
load_dotenv()
enable_post = bool(os.getenv("ENABLE_POST"))
ifttt_key = str(os.getenv("IFTTT_KEY"))
motion_detected_cooldown = int(os.getenv("MOTION_DETECTED_COOLDOWN"))
read_frequency = float(os.getenv("READ_FREQUENCY"))
notification_frequency = int(os.getenv("NOTIFICATION_FREQUENCY"))

print("enable_post loaded as: " + str(enable_post))
print("motion_detected_cooldown loaded as: " + str(motion_detected_cooldown))
print("read_frequency loaded as: " + str(read_frequency))
print("notification_frequency loaded as: " + str(notification_frequency))

# Instantiate Program Vars
beganCountDateTime = datetime.now()
beganCountTimeStamp = time.time()
motionDetectedCount = 0
lastOccurenceTimeStamp = 0

beganCountDateTimeFormatted = beganCountDateTime.strftime("%d/%m/%Y %H:%M:%S")
print("Current Date/Time: " + beganCountDateTimeFormatted)

try:
	print("Waiting for PIR to settle ...")
	
	# Loop until PIR output is 0
	while GPIO.input(pinpir) == 1:
	
		currentstate = 0

	print("    Ready. Quit at any time with CTRL-C")
	# Sleep momentarily to allow messages to be read
	time.sleep(1.5)
	
	# Loop until users quits with CTRL-C
	while True:
	
		# Read PIR state
		currentstate = GPIO.input(pinpir)

		# If the PIR is triggered
		if currentstate == 1 and previousstate == 0:
		
			# Make a note
			motionDetectedCount += 1

			currentTimeStamp = time.time()

			# See how long it has been since we started counting
			if lastOccurenceTimeStamp > 0:
				sinceLastOccurenceSecs = currentTimeStamp - lastOccurenceTimeStamp
				sinceLastOccurenceMins = round(sinceLastOccurenceSecs / 60, 2)
				sinceBeganCount = currentTimeStamp - beganCountTimeStamp

				print(
					"Motion detected! " + str(sinceLastOccurenceMins) + 
					" minutes elapsed since the last occurence. Total since last reset: " + str(motionDetectedCount)
				)

				# If we have passed the notification threshold then send an update if enabled
				if sinceBeganCount > notification_frequency:
					if enable_post:
						print("Post enabled - pushing event to IFTTT")

						sinceBeganCountHours = round((sinceBeganCount / 60) / 60, 2)

						# Your IFTTT URL with event name, key and json parameters (values)
						r = requests.post(
							'https://maker.ifttt.com/trigger/motion_detected/with/key/' + ifttt_key,
							params={"value1":motionDetectedCount,"value2":sinceBeganCountHours,"value3":"none"}
						)
					else:
						print("Post disabled - nothing sent")

					# Reset count
					beganCountTimeStamp = time.time()
					motionDetectedCount = 0
			else:
				print("Motion detected! The first since the last reset")
			
			# Record new previous state
			previousstate = 1
			lastOccurenceTimeStamp = currentTimeStamp
			
			#Wait the specified number of seconds before looping again
			print("    Waiting " + str(motion_detected_cooldown) + " seconds")
			time.sleep(motion_detected_cooldown)
			
		# If the PIR has returned to ready state
		elif currentstate == 0 and previousstate == 1:
		
			print("    Resetting")
			previousstate = 0

		# Wait for the specified number of seconds
		time.sleep(read_frequency)
		print("    Ready")

except KeyboardInterrupt:
	print("    Quit")

	# Reset GPIO settings
	GPIO.cleanup()

start.py
#! /usr/bin/python

# Based on https://cdn.shopify.com/s/files/1/0176/3274/files/iftttpir.py from the PiHut

# Imports
import RPi.GPIO as GPIO
import time
import requests
from dotenv import load_dotenv
import os
from datetime import datetime

# Set the GPIO naming convention
GPIO.setmode(GPIO.BCM)
# Turn off GPIO warnings
GPIO.setwarnings(False)
# Set a variable to hold the GPIO Pin identity
pinpir = 17
# Set GPIO pin as input
GPIO.setup(pinpir, GPIO.IN)

# Variables to hold the current and last states
currentstate = 0
previousstate = 0

# Load Environment Variables
load_dotenv()
enable_post = bool(os.getenv("ENABLE_POST"))
ifttt_key = str(os.getenv("IFTTT_KEY"))
motion_detected_cooldown = int(os.getenv("MOTION_DETECTED_COOLDOWN"))
read_frequency = float(os.getenv("READ_FREQUENCY"))
notification_frequency = int(os.getenv("NOTIFICATION_FREQUENCY"))

print("enable_post loaded as: " + str(enable_post))
print("motion_detected_cooldown loaded as: " + str(motion_detected_cooldown))
print("read_frequency loaded as: " + str(read_frequency))
print("notification_frequency loaded as: " + str(notification_frequency))
time.sleep(1)

# Instantiate Program Vars
beganCountDateTime = datetime.now()
beganCountTimeStamp = time.time()
motionDetectedCount = 0
lastOccurenceTimeStamp = 0
connectionTimeout = 60 # How many seconds to wait for IFTTT to respond in secs
retryPostCounter = 0 # Tracks how many attempts to post before giving up

beganCountDateTimeFormatted = beganCountDateTime.strftime("%d/%m/%Y %H:%M:%S")
print("Current Date/Time: " + beganCountDateTimeFormatted)

try:
	print("Waiting for PIR to settle ...")
	
	# Loop until PIR output is 0
	while GPIO.input(pinpir) == 1:
	
		currentstate = 0

	print("    Ready. Quit at any time with CTRL-C")
	# Sleep momentarily to allow messages to be read
	time.sleep(1.5)
	
	# Loop until users quits with CTRL-C
	while True:
	
		# Read PIR state
		currentstate = GPIO.input(pinpir)

		# If the PIR is triggered
		if currentstate == 1 and previousstate == 0:
		
			# Make a note
			motionDetectedCount += 1

			currentTimeStamp = time.time()

			# See how long it has been since we started counting
			if lastOccurenceTimeStamp > 0:
				sinceLastOccurenceSecs = currentTimeStamp - lastOccurenceTimeStamp
				sinceLastOccurenceMins = round(sinceLastOccurenceSecs / 60, 2)
				sinceBeganCount = currentTimeStamp - beganCountTimeStamp

				print(
					"Motion detected! " + str(sinceLastOccurenceMins) + 
					" minutes elapsed since the last occurence. Total since last reset: " + str(motionDetectedCount)
				)

				# If we have passed the notification threshold then send an update if enabled
				if sinceBeganCount > notification_frequency:
					if enable_post:
						print("Post enabled - pushing event to IFTTT")

						sinceBeganCountHours = round((sinceBeganCount / 60) / 60, 2)

						try:
							# Your IFTTT URL with event name, key and json parameters (values)
							# Timeout is how long to wait in seconds in the event of no response or internet dropping
							r = requests.post(
								'https://maker.ifttt.com/trigger/motion_detected/with/key/' + ifttt_key,
								params={"value1":motionDetectedCount,"value2":sinceBeganCountHours,"value3":"none"},
								timeout=connectionTimeout
							)

							retryPostCounter = 0
						except requests.exceptions.ConnectionError:
							retryPostCounter += 1

							if retryPostCounter >= 5:
								raise ValueError("Attempts to contact IFTTT reached threshold. Aborting :(")
							else:
								print("Error pushing to IFTTT - ConnectionError, will try again...")
								time.sleep(connectionTimeout)
						except requests.Timeout:
							retryPostCounter += 1

							if retryPostCounter >= 5:
								raise ValueError("Attempts to contact IFTTT reached threshold. Aborting :(")
							else:
								print("Error pushing to IFTTT - Timeout, will try again...")
								time.sleep(connectionTimeout)
					else:
						print("Post disabled - nothing sent")

					# Reset count
					beganCountTimeStamp = time.time()
					motionDetectedCount = 0
			else:
				print("Motion detected! The first since the last reset")
			
			# Record new previous state
			previousstate = 1
			lastOccurenceTimeStamp = currentTimeStamp
			
			#Wait the specified number of seconds before looping again
			print("    Waiting " + str(motion_detected_cooldown) + " seconds")
			time.sleep(motion_detected_cooldown)
			
		# If the PIR has returned to ready state
		elif currentstate == 0 and previousstate == 1:
		
			print("    Resetting")
			previousstate = 0

		# Wait for the specified number of seconds
		time.sleep(read_frequency)
		print("    Ready")

except KeyboardInterrupt:
	print("    Quit")

	# Reset GPIO settings
	GPIO.cleanup()